### ISSUE TITLE
False positive E0107 with std::sync::Weak::upgrade and std::io::Write

### ISSUE THREAD

**Issue by Meralis40** [rust-lang/rust-analyzer#22297]
rust-analyzer resolves incorrectly types when using `std::sync::Weak::upgrade` (the resulting type is shown as `Arc<{unknown}>`. This cause E0107 when calling `upgrade()` on a `Weak<RwLock<T>>` and `std::io::Write` is in the scope, and `RwLock::write()` is called on the upgraded `Arc`.
This does not happen with `std::rc::Weak`.
In both case, `cargo check` returns without any error.

**rust-analyzer version**: rust-analyzer version: 0.3.2887-standalone (f04c372864 2026-05-03)

**rustc version**: rustc 1.97.0-nightly (e95e73209 2026-05-05)

**editor or extension**: VSCode v1.118.1 + rust-analyzer v0.3.2887

**relevant settings**: rust-analyzer extension settings:
```json
{
    "rust-analyzer.cargo.targetDir": "${workspaceFolder}/target/rust-analyzer",
    "rust-analyzer.hover.actions.references.enable": true,
    "rust-analyzer.restartServerOnConfigChange": true,
    "rust-analyzer.testExplorer": true,
    "rust-analyzer.assist.emitMustUse": true,
    "rust-analyzer.server.extraEnv": {
        "CMAKE_GENERATOR": "Ninja"
    },
    "rust-analyzer.cargo.extraEnv": {
        "CMAKE_GENERATOR": "Ninja"
    }
}
```

**code snippet to reproduce**:
```rust
use std::sync::Weak;
// use std::rc::Weak;
use std::io::Write;
use std::sync::RwLock;

fn main() {
    let t : Weak<RwLock<()>> = Weak::new();
    if let Some(x) = t.upgrade() {
        // r-a shows error E0107 here (expected 1 argument, found 0)
        // If std::rc::Weak is used instead, r-a shows no error
        let _unused = x.write();
    }
}
```

**Comment by ChayimFriedman2** [comment]
I don't think this is a duplicate of https://github.com/rust-lang/rust-analyzer/issues/22171, but I can't reproduce with nightly r-a.

**Comment by A4-Tacks** [comment]
I think it's a duplicated because I can reproduce it in latest nightly (0.4.2890-standalone), and after removing `pattern_type` from the core source `NonNull`, it works

rustc 1.97.0-nightly (36ba2c771 2026-04-23)

**Comment by A4-Tacks** [comment]
I think the difference between `Rc` and `Arc` is that there is `impl Write for Arc<File>` but no `impl Write for Rc<File>`

**Comment by ChayimFriedman2** [comment]
Hmm, I can reproduce this but only on nightly toolchain. I also see "type annotations needed" error, even with `rc::Weak`.

**Comment by ChayimFriedman2** [comment]
Yes, stable does not use `pattern_type!()`. I think it's because we need the type of the last field to calculate the sizedness of the struct.

**Comment by A4-Tacks** [comment]
> I also see "type annotations needed" error

Because as stated in #22171, the type is `{unknown}`
